### ISSUE TITLE
Redesign: Register Field Type vs. Instance of Type

### ISSUE THREAD

**Issue by joemcbride** [joemcbride/graphql-dotnet#3]
Now that I understand GraphQL a lot better, one of the flaws of the API I created is the registration of an instance of the field Type vs. just the Type itself.  This has lead to issues with completing the introspection feature.

To fix this I think all field registrations will change to something like:

```
Field<StringGraphType>("name", "The name of the person.");
```

This syntax is currently supported, though it immediately creates an instance of the given type.  This will be changed to just capture the type and the instance of the type will be created later.

There really only ever needs to be a single instance of each defined type.  At runtime a type will be requested and it will check a cache to see if an instance of that type has been created yet.  This is sort of already being done within the [Schema class](https://github.com/joemcbride/graphql-dotnet/blob/master/src/GraphQL/Types/Schema.cs#L13-L18).  The difference will be the types will be created here vs. just collecting them into a lookup.  I will design this so that if you want to resolve the type through an IOC container you can.

**Comment by joemcbride** [comment]
Closing this.  Design issue resolved with the 3.0 release.  https://github.com/joemcbride/graphql-dotnet/releases/tag/v0.3.0

**Comment by plecong** [comment]
Hi @joemcbride, 

First off, great work here and I'm super excited about this project. 

**TL;DR** The current use of .NET Types as GraphQL types is conflating concepts and could lead to some problems down the road.

I realize this issue is closed, but having played with the graphql-js implementation I was puzzled by the change from "instance of GraphQL type" to "Type of GraphQL type". I actually think this wasn't a flaw in the original API design.

While I see that you can now use a GraphQL type Type class as the generic parameter, I'm not sure it buys you anything. The generic parameter is for .NET, and not for GraphQL's type system. What I mean is that users of the API aren't benefiting from knowing that the generic type is the GraphQL type.

Both of these examples are equivalent:
```csharp
// Now: Type as GraphQL type
Field<StringGraphType>("name", "The name of the person")

// Before: Instance as GraphQL type
Field(Scalars.GraphQLStringType, "name", "The name of the person")
```

However, with method generic type information which .NET knows how to use, you could define the `Field()` DSL method on `GraphType` like:

```csharp
public void Field<TResolve>(
    GraphType type, 
    string name,
    string description = null,
    QueryArguments arguments = null,
    Func<ResolveFieldContext, TResolve> resolve = null) 
```

This enables type safety when creating the schema:

```csharp
Field<string>(
    Scalars.GraphQLString, 
    "name", 
    description: "The name of the person",
    resolve: ctx => "John"); // <- this must return a string
```

I was taking a look at converting the Union and other missing types, but things get a little mixed up. For example, a GraphQLUnionType consists of two or more other GraphQL types, so one way to do this is to make the union of types a generic parameters on the `UnionGraphType`:

```csharp
public class UnionGraphType<T1,T2> : GraphType 
    where T1 : GraphType 
    where T2 : GraphType { }
```

This would allow the `Field` DSL method to look like this:

```csharp
Field<UnionGraphType<FooType,BarType>>("FooBar", "A foo or a bar");
```

Unfortunately, this can get unwieldly and we'd have to limit the number of sub-types in the GraphQL union (which probably isn't bad).

Using .NET Types as GraphQL types also means that we have to instantiate the type to get properties on it because the .NET Type doesn't give a way to get additional information through Reflection (except through Attributes). This leads to the need for `GraphTypesLookup` and `TypeCollectionContext`.

All this being said, I believe the reason (and correct me if I'm wrong) for changing to this was so that the Schema DSL had .NET "feel" where code liked `Field<StringGraphType>` read like, "Field of this GraphQL String type". Unfortunately this is conflating the GraphQL type (which graphql-js represents as an instance) with the .NET type.

Also, looking down the road, I saw there was an issue for auto-generating schema. I think having GraphQL types as .NET types would ultimately make this very complicated. For example, from #14 we have the following:

```csharp
    public abstract class StarWarsCharacter {
        [NotNull]
        [Description("The id of the character.")]
        public int Id { get; set; }

        [Description("The name of the character.")]
        public string Name { get; set; }

        // If the function returns a Task, the executor will execute it asynchronously
        public async Task<IList<StarWarsCharacter>> Friends() {
            return await _Repository.GetFriendsForIdAsync(Id);
        }

        [Description("Which movie they appear in.")]
        public async Task<IList<Episode>> AppearsIn() {
            return await _Repository.GetAppearsInById(Id);
        }
    }
```

To make this work, there would have to be code generation to create a .NET Type to represent it in the schema:

```csharp
// This would have to be generated or manually created as boilerplate.
public class StarWarsCharacter : GraphObjectType { }
```

However, if you had instances of GraphQL types you could potentially build an instance by Reflecting on the properties and attributes:

```csharp
public static class GraphQLTypeExtensions
{
    public static ObjectGraphType AsGraphType(this Type type)
    {
        // scan through properties and create fields automatically using 
        // attributes for name/descriptions
    }
}
```

Then in code your schema DSL would be:

```csharp
var starWarsCharacter = typeof(StarWarsCharacter).AsGraphType();
```

I realize there was a lot of work put into this to get it down this road, but I'm a little concerned that this disconnect with the reference implementation might impede full implementation down the road.

**Comment by joemcbride** [comment]
Hey, thanks for the feedback.  I'm traveling this weekend though can write a proper response on Monday.

**Comment by joemcbride** [comment]
Again, thanks for your feedback.  I was hoping once this was out there someone would come along and say "hey, this this has some issues, maybe we can do it better?".  Let me see if I can explain some of my original reasoning and hopefully we can come up with better ways to do it.

**Requirements**

1. There only ever needs to be a single instance of any given "base" type in the entire schema.  A base type would be scalar types and user-defined types.

2. Collecting distinct type data for introspection.  For introspection to work, you need to know what distinct low-level types are available in your system.  In addition, you need to know what types "implement" which interfaces.

3.  Resolving dependencies through a IOC container.  You may have a service that needs to fetch data in your `resolve` method.  Typically you would inject this dependency in the constructor of your class.

For req 1, this is a little easier in JavaScript since making globals is "on by default", and you can easily import instances of classes from other files.

For req 2, this is where the `GraphTypesLookup` originally came into play in conjunction with req 1.  Have a place for the code to just say "give me this type" and it hands it over.

For req 3, I was originally thinking of just constructor injection.  Thinking more on this, it probably would have been better to go with a service location pattern, adding  a `GetService<T>` on the `ResolveFieldContext`.

The change to Types vs. instances lead me to `TypeCollectionContext`, [which I really don't like](https://github.com/joemcbride/graphql-dotnet/blob/master/src/GraphQL/Types/GraphType.cs#L53).  I would be happy to get rid of it.

**Dynamic DSL**
For this to work I think you would create an object and pre-register it with the type lookup.  Creating a generic type in code is pretty easy, so I think we would need to add an overload for field registration that took just the `Type`.  Having all instances could maybe make this a little easier, though doing it with with Types shouldn't be terribly hard.  I think we're talking about a difference of `Field(obj, "name")` vs. `Field(obj.GetType(), "name")`.

**Unions**

For unions specifically, I probably wouldn't suggest a `UnionGraphType<T1, T2>`.  This is how they do it in the JavaScript project:

```
var PetType = new GraphQLUnionType({
  name: 'Pet',
  types: [ DogType, CatType ],
  resolveType(value) {
    if (value instanceof Dog) {
      return DogType;
    }
    if (value instanceof Cat) {
      return CatType;
    }
  }
});
```

In .NET, with the current design of GraphQL.NET I would probably do something like:

```
public class PetType : UnionGraphType
{
  public PetType()
  {
    Types.Add(typeof(DogType));
    Types.Add(typeof(CatType));
  }
  ...
}
```

Then when you register the field:

```
Field<ListGraphType<PetType>>("pets");
```

**A Better Way**

* Use service location with `ResolveFieldContext`.
* I like the idea of having a strongly typed result on field resolution, will take some updates to `FieldType` (putting in a generic and non-generic interface I think), though can probably do that.
* What pattern do you see being used to resolve req 1 with user-created types?  Have a "schema builder" class where the instances are contained and passed around?  Static instances on the type itself like I had on the Scalar types?  Static instances on the Schema? Schema builder + type builder?  I'm going to take a look at the Java project and see if I can discern how they went about doing this.

```
public class SchemaBuilder
{
  public Schema BuildSchema()
  {
    var humanType = new HumanType(...);
    var humanType = ObjectGraphType.Build("Human", ...);

    var query = new QueryType(humanType);

    return new Schema
    {
      Query = query
    };
  }
}
```

```
public class SomeType : ObjectGraphType
{
  public static SomeType Instance = new SomeType();
  protected SomeType()
  {
    ...
  }
}
```

**Comment by joemcbride** [comment]
@moserware @scmccart Feedback welcome ^^

**Comment by plecong** [comment]
Hi,

Thanks for the reply and I do think there's ways to solve the problems you state. 

**Requirement 1**

So, the way that the reference implementation solves this is that they create a `__Schema` singleton instance that represents the "base" types. I would imagine that a static variable on a static class like `Introspection` would suffice to contain this as a global in .NET.

This is essentially the same as having a schema builder instance, but the `__Schema` instance is more likely an internal singleton.

**Requirement 2**

That `__Schema` global is used in their `Schema` object during the type discovery phase (let's call it). A similar technique now that the object is global can be used in .NET. This essentially merges all of the "base" and "user-defined" types together into the single schema instance.

I'm also thinking that the Schema would also be an instance. The instance could be of a type that inherits from Schema or just be an single value constructed or just a singleton pattern:

```csharp
public class StarWarsSchema extends Schema 
{
    public static readonly StarWarsSchema INSTANCE = new StarWarsSchema();

    public StarWarsSchema()
    {
        Query = // build your query schema here. 
    }
}
```

**Requirement 3**

So using IoC with an instance type where the .NET type is the generic type instead of the GraphQL type actually makes more sense to me.

For instance, let's say you have a `Foo` domain class:

```csharp
public class Foo 
{
    public string Name { get; set; }
}
```

And with that you have something that "loads" the Foo from a database or REST call:

```csharp
public interface ILoader<T>
{
    T GetItem(int id);
}
```

And now you create a GraphQL type that represents `Foo` in GraphQL:

```csharp
public class FooSchema : GraphQLSchema
{
    public FooSchema(ILoader<Foo> fooLoader)
    {
        var fooType = new GraphQLObjectType<Foo>(
            name: "Foo",
            fields: builder => {
                builder.Field("name", GraphQLScalars.GraphQLString, f => f.Name);
            });

        Query = new GraphQLObjectType(
            name: "Query",
            fields: builder =>
            {
                builder.Field<Foo>("foo", resolve: (context) => fooLoader.GetItem(context.args["id"]));
            });
        }
    }
}
```

This could get unwieldly when there are a lot of different loaders and types. The pattern that I think would be helpful then is to create provider objects that create GraphQL type instances for related subsets of types that are fetched together. Then the top-level Schema would compose these parts together like modules that are provided by the IoC container.

**Other Considerations**

I've been playing with how this would work with various builder patterns and one tricky problem that arises with instance vs Type is the circular reference problem. The Java project solved this by introducing a "type reference" type. During the "type discovery" phase, the references would be resolved to their real instance. There would have to be some validation to ensure that there aren't leftover references and it makes schema construction a little more difficult.

Also, I've been playing with a prototype and using the generic type parameter as a domain type instead of a GraphQL type also enables really easy schema creation using Reflection:

```csharp
// with the right defaults this is all that is really needed
// because the name and fields can be inferred.
var fooType = new GraphQLObjectType<Foo>();
```

That has opened up the builder API I'm toying with because then you could create predicates on which properties you want to expose, have the builder just use a Linq Expression to generate the resolve and name, etc.

```csharp
var fooType = new GraphQLObjectType<Foo>(fields: builder => {
    // as an Expression I could get the name of the property and have the resolve built-in
    builder.Field(f => f.Name);
});
```

This could make schema creation extremely easy for existing .NET projects.

**Comment by joemcbride** [comment]
I'm liking that syntax.

How would the field

```
builder.Field<Foo>("foo", resolve: (context) => fooLoader.GetItem(context.args["id"]));
```

get resolved to the `fooType`?  Would that be

```
builder.Field<Foo>("foo", fooType, resolve: (context) => fooLoader.GetItem(context.args["id"]));
```

instead?

**Comment by joemcbride** [comment]
List/NonNull syntax:

```
fields: _ =>
{
    _.Field("name", new GraphQLNonNull(GraphQLScalars.GraphQLString), f => f.Name);
    _.Field("friends", new GraphQLList(GraphQLScalars.GraphQLString), f => f.Friends);
});
```

**Comment by plecong** [comment]
That's correct. The `Field` method has some overloads to be able to work correctly.

Here's what I'm playing with:

```csharp
public class GraphQLObjectType<T> :
    GraphQLObjectType,
    IGraphQLType,
    IGraphQLOutputType,
    IGraphQLCompositeType,
    IGraphQLNullableType,
    IGraphQLNamedType
{
    public GraphQLObjectType(
        string name,
        IEnumerable<GraphQLInterfaceType> interfaces = null,
        Action<IGraphQLFieldBuilder<T>> fields = null,
        Func<object, bool> isTypeOf = null,
        string description = null) : base(name, interfaces, isTypeOf: isTypeOf, description: description)
    {

    }        
}   

public interface IGraphQLFieldBuilder<T>
{
    void Field(
        string name,
        IGraphQLOutputType type = null,
        string description = null,
        IGraphQLFieldResolver<T> resolver = null);

    void Field<TResolve>(
        string name,
        IGraphQLOutputType type = null,
        string description = null,
        IGraphQLFieldResolver<T,TResolve> resolver = null);

    void Field<TResolve>(
        string name,
        IGraphQLOutputType type = null,
        string description = null,
        Func<T, GraphQLResolveInfo, TResolve> resolve = null);

    void Field<TResolve>(
        string name,
        IGraphQLOutputType type = null,
        string description = null,
        Func<T, TResolve> resolve = null);
}
```

The `IGraphQLFieldResolver` is what is responsible for resolving the field from type `T` to type `TResolve`. It's possible to create a re-direction around an existing `Func<>` or interfaces that are responsible for the loading.

What I'm finding is that the generic and non-generic versions of the constructors can get unwieldly. I'm thinking that it's possible to expose two different APIs that are largely similar, but would make the constructors more consistent. Essentially, where there are a lot of overloads due to options for optional parameters, we would expose that as method overrides. Then for classes, expose a more traditional Builder pattern (e.g. inner-class, setter methods, etc.).

The `IGraphQLFieldBuilder` above illustrates creating the method overloading pattern. It also shows how the constructor for `GraphQLObjectType` can get pretty hairy with generic parameters. Instead, we can expose methods on `GraphQLSchema`:

```csharp
public class GraphQLSchema {

    protected GraphQLObjectType ObjectType(...);
    protected GraphQLListType ListType(...);
    protected GraphQLNonNull NonNull(...);
    protected GraphQLTypeReference TypeRef(...);

}
```

This gives consumers of the API that are building the schema built-in helper methods when creating a schema for 80% options. Or they can go to the more advanced "kitchen-sink" Builder for fully customizing the GraphQL types.

We may want to expose methods like `MapType` on `GraphQLSchema` that do the Reflection based schema constructions so developers have a scalable API surface area from easy (Reflection-based, get it up and working fast on existing models), to medium (add some descriptions and get some depth), to full-blown.

**Comment by plecong** [comment]
I'm probably over thinking this, but I'm trying to think through the use-cases for the API. For the IoC use-case it seems like there would be a need for a "bridging" class that could be created.

Perhaps, the helper methods above could be placed on an abstract class like `SchemaModule`:

```csharp

public class Page { }
public class Post
{
    public int Id { get; set; }
}
public class Comment { }

public interface ICommentLoader
{
    IEnumerable<Comment> GetComments(int postId);
}

public interface IPostLoader
{
    IEnumerable<Post> GetPosts();
    Post GetPost(int id);
}

public class PageSchemaModule : GraphQLSchemaModule { }

public class PostSchemaModule : GraphQLSchemaModule
{
    public GraphQLObjectType PostType { get; private set; }

    public GraphQLObjectType CommentType { get; private set; }

    public PostSchemaModule(ICommentLoader loader)
    {
        CommentType = new GraphQLObjectType<Comment>("Comment");

        PostType = new GraphQLObjectType<Post>(
            name: "Post",
            fields: _ =>
            {
                _.Field(
                    name: "Comments", 
                    type: new GraphQLList(CommentType), 
                    resolve: (post, context) => loader.GetComments(post.Id));
            });
    }

    public IEnumerable<Post> GetPosts() { }

}

public class WordpressSchema : GraphQLSchema
{
    public WordpressSchema(
        PostSchemaModule postModule,
        PageSchemaModule pageModule)
    {
        Query = new GraphQLObjectType(
            name: "RootQuery",
            fields: _ =>
            {
                _.Field(
                    name: "posts", 
                    type: new GraphQLList(postModule.PostType), 
                    resolve: ctx => postModule.GetPosts());
            });
    }
}
```

The IoC container can then build up the modules using constructor injection and output a schema. The module is a way to decompose the schema into sections or even aggregate roots (if you speak DDD).

**Comment by joemcbride** [comment]
I'm not an expert on DDD though I understand the aggregate concept.  That pattern would certainly resolve the IoC issues I was thinking of.  It also allows you to split up the schema definition so you don't have one massive definition in a single file.

I like the idea of having simplified registration to get going quickly, which then you could build upon.

`IGraphQLFieldResolver` - perfect.

I setup that sample code locally and played with it a bit, really liking it.  I think the one suggestion I have so far is to make the `fields` required.  I don't think creating a `GraphQLObjectType` without fields is valid.

As a side note, I also think it is probably a good idea to switch to the same names as the reference implementation, as you have done.  Makes looking at examples easier to understand I think as well as comparing to the reference implementation code.  That will be unfortunate for anyone who has used the project so far (which isn't huge), though shouldn't be terrible.  Better to do that now rather than later.

**Comment by plecong** [comment]
Re: Making `fields` required, I was going back and forth on this and I think you're right. My original thinking was that without a fields definition that it would automatically reflect, but I think for constructors being explicit is a better API. 

I was toying with a `FieldsFromReflection` method on the `IGraphQLFieldBuilder`:

```csharp
            var humanType = new GraphQLObjectType<Human>(
                name: "Human",
                description: "A humanoid creature in the Star Wars universe.",
                fields: _ => _.FieldsFromReflection());
```

I'm also going back on forth on if there should be a `IGraphQLFieldsBuilder` for building `fields` on a a `GraphQLObjectType` and when you call `Build()` or `BuildField()` on it you get an `IGraphQLFieldBuilder`:

```csharp
            var humanType = new GraphQLObjectType<Human>(
                name: "Human",
                description: "A humanoid creature in the Star Wars universe.",
                fields: _ =>
                {
                    _.BuildField("id")
                        .Description("The id of the character")
                        .Resolve(x => x.Id);
                });
```

Re: Using the same names as the reference implementation, again I was going back and forth. I originally setup my model like theirs so that I could write samples that looked similar. I found the names a tad verbose, but having sample code that looks like the reference is probably beneficial in the end.
